### ISSUE TITLE
Finding locations of text and "porting" to it

### ISSUE THREAD

**Issue by robmburke** [jacobslusser/ScintillaNET#160]
I am trying to create a window that will grab the text from every file in the project and find the "TODO:" comments.  Here is the code:

    //Method on ToDo Form to populate locations and store the expected cursor positions
	List<ToDoMatch> matches = new List<ToDoMatch>();

	foreach (FileInfo file in Files)
	{
		using (Scintilla scintilla = new Scintilla())
		{
			using (StreamReader reader = new StreamReader(file.FullName))
			{
				scintilla.Text = reader.ReadToEnd();
			}

			int result = 0;
			scintilla.CurrentPosition = 0;

			do
			{
				scintilla.TargetStart = scintilla.CurrentPosition;
				scintilla.TargetEnd = scintilla.TextLength;

				result = scintilla.SearchInTarget("TODO:");

				if (result != -1)
				{
					scintilla.SetSel(scintilla.TargetStart, scintilla.TargetEnd);

					Line line = scintilla.Lines[scintilla.CurrentLine];

					matches.Add(new ToDoMatch()
					{
						File = file,
						Source = file.Name,
						LineText = string.Format("Line {0}: {1}", line.Index + 1, line.Text),
						Position = scintilla.TargetStart
					});
				}
			} while (result != -1);
		};
	}

    //Method on Main Form to open editor, and place the cursor
	public void DisplayToDo(ToDoLocater locater)
	{
		DisplayEditor(locater.File);
		
		Scintilla scintilla = (Scintilla)workspace.SelectedTab.Controls[1];
		scintilla.Select();
		scintilla.Focus();

		scintilla.AnchorPosition = locater.Position;
		scintilla.CurrentPosition = locater.Position;
	}

The problem is that when I use the TargetStart value to try to put the cursor at that position, it's off by like 2 spaces past the position I am expecting for the first match.  Subsequent, matches in the same file are an extra 2 spaces from the expected starting point.

Basically, it seems my search method above seems to be reporting a number that is different once I actually do open a tab with an editor in it and try to place it at the number I am given; it doesn't match up with what is on the screen

**Comment by jacobslusser** [comment]
That's interesting. I'm not sure I know for sure what could be the cause, but I have a hunch if you are interested in exploring it. Do your files have a [Byte Order Mark (BOM)](https://en.wikipedia.org/wiki/Byte_order_mark)? If so, try removing it by saving your files in an editor that allows you to save without a BOM or use a Hex Editor to remove it and tell me if that makes a difference on those files.

**Comment by robmburke** [comment]
Haha...I'm an idiot.
My "IDE" is a special IDE that does text conversions on save.
So the text you see is different than the text you save based on some variables the user can define.

I need to do those conversions when loading the text into this "virtual" scintilla control as well.  Sorry for this.

**Comment by JohnyMac** [comment]
Just a tip, why not use regex ?
You are just looking to parse for TODO's right ? Regex will do just better.

**Comment by robmburke** [comment]
OK, So like I said, I got the original issue fixed.

Along the same lines of "porting" to the shown text, I didn't originally test locations off the screen.  I double-click the ListViewItem, it moves the cursor to the appropriate location but the visible lines don't change.  I sort of assumed this would be the default since you usually want to see where your cursor so didn't think twice about needing to code for it.

How do you force the control to show the line that the cursor is currently at?

**Comment by robmburke** [comment]
@JohnyMac I thought about it, but I used the scintilla control because if I just use a basic string what about characters that are represented by more than 1 byte?  Wouldn't the positional values be different than if I used the built-in Scintilla functionality that makes allowance for those types of characters?  I'm not anticipating character sets like that because this is just a simple SQL editor.  But I just wanted there to not be any mistakes with the presentation.

Are you saying to just .NET regex a string?  or still create a "virtual" scintilla control to be able to parse the text with scintilla and use the built-in scintilla regex?

**Comment by jacobslusser** [comment]
`Scintilla.ScrollCaret()` will scroll the caret into view.

**Comment by JohnyMac** [comment]
In my project, I just get all the text using the normal Text property and then parse it using regex and for performance, I only parse visible parts of code on-runtime.
I don't store anything about their positions but I will need them in the future and you just got my mind about it.. So no idea how I would handle it to be honest... I will probably just use the normal search in ScintillaNET for getting the positions on runtime when needed, since I need the best possible speed in the parser.

For reference, Here is where I have my parser: https://github.com/JohnyMac/ExtremeStudio/tree/master/ExtremeParser
*(The main file is called `Parser.vb` and you can find all the classes in the folder called `Parsers`)*

And where I actually call it: https://github.com/JohnyMac/ExtremeStudio/blob/master/ExtremeStudio/DockPanel%20Forms/MainForm%20Docks/EditorDock/EditorDock.vb#L153-L421
*(Its a bit complicated to explain but that's the important part)*

Soo, IDK which way is more efficient to be honest but that's what I do.
